### PR TITLE
Standard CRC32 table based algorithm implemented.

### PR DESCRIPTION
**Pull request by CLomanno**
I had to redefine the polys and expected CRCs for everything in CRC32.  However I modified the functions in crc32 so they are all backwards compatible.

The new function `new_with_initial_and_final()` allows you to generate a reflected table with a custom initial and final XOR value.

@mrhooray 
Sorry, this is my first pull request and I didn't add a reviewer, assignee, or label.

### CODE REVIEW

**Comment by CLomanno** [comment]
I found the default values for various 32bit crc standards here: https://en.wikipedia.org/wiki/Cyclic_redundancy_check#Polynomial_representations_of_cyclic_redundancy_checks

Verified the correct crc values here: http://crccalc.com/

**Comment by CLomanno** [comment]
@mrhooray 
Sorry, this is my first pull request and I didn't add a reviewer, assignee, or label.

**Comment by CLomanno** [comment]
@mrhooray I completed updating all 3 CRC types.  Now crc16, crc32, and crc64 are backwards compatible and can work normal or reflected.

File: Cargo.toml [hunk 1]
```diff
@@ -12,6 +12,13 @@ license = "MIT OR Apache-2.0"
 [build-dependencies]
 build_const = "0.2"
 
+#[dev-dependencies]
```
Comment (mrhooray) [inline Cargo.toml]: shall we remove unrelated changes or you want to move benchmarks to use criterion instead in this pr?

File: benches/bench.rs [hunk 2]
```diff
@@ -7,24 +7,65 @@ use test::Bencher;
...
 fn bench_crc64_update_megabytes(b: &mut Bencher) {
-    let table = crc64::make_table(crc64::ECMA);
+    let table = crc64::make_table(crc64::ECMA, true);
     let bytes = Box::new([0u8; 1_000_000]);
-    b.iter(|| crc64::update(0, &table, &*bytes));
+    b.iter(|| crc64::update(0, &table, &*bytes, true));
 }
+
+/*
+#[macro_use]
```
Comment (mrhooray) [inline benches/bench.rs]: ditto

File: src/crc16.rs [hunk 3]
```diff
@@ -1,16 +1,31 @@
...
 
 include!(concat!(env!("OUT_DIR"), "/crc16_constants.rs"));
 
+/// Structure that holds all of the important values for calculating a CRC
+///
+/// # Definitions
+///
+/// *table:* Holds the table values based on the supplied polynomial for the fast CRC calculations
+///
+/// *initial:* The initial inut value. AKA reflect_in
```
Comment (mrhooray) [inline src/crc16.rs]: nit: `input` instead of `inut`, and a few other occurrences in this pr

File: Cargo.toml (hunk 1 above)
Comment (CLomanno) [reply Cargo.toml]: Criteron is nice, but we can use whatever you want.  I left both in for you to choose

File: src/crc16.rs (hunk 3 above)
Comment (CLomanno) [reply src/crc16.rs]: A lot of the comments were copy/pasted between crc16/32/64

File: Cargo.toml (hunk 1 above)
Comment (mrhooray) [reply Cargo.toml]: cool, let's remove the commented code and revisit benchmark framework in separate pr?

File: src/crc16.rs (hunk 3 above)
Comment (mrhooray) [reply src/crc16.rs]: do you want to keep it, or it's actually a typo?

File: src/util.rs [hunk 4]
```diff
@@ -1,47 +1,191 @@
...
+        if (value & 0x01) == 1 {
+            reflection |= 1 << ((bits - 1) - i)
+        }
+        value = value >> 1;
+    }
+    reflection
+}
+
+/// Reflects the lease significant byte of a u32.
+pub fn reflect_long_32(input: u32) -> u32 {
```
Comment (mrhooray) [inline src/util.rs]: have following 3 functions named consistently like `reflect_byte_16/32/64`?

File: src/util.rs [hunk 5]
```diff
@@ -1,47 +1,191 @@
...
+            value = reflect_long_long(value);
+        }
+
+        table[i as usize] = value;
     }
     table
 }
+
+/// Reflects a value of a 16 bit number
+pub fn reflect_short(mut value: u16) -> u16 {
```
Comment (mrhooray) [inline src/util.rs]: `reflect_16/32/64` or something similar?
just to make things consistent with next set of 3 functions

**Comment by nickbabcock** [comment]
As a side note, I benchmarked this impl against:

```rust
pub fn calc_crc(data: &[u8]) -> u32 {
    !data.iter().fold(!0xefcb_f201, |acc, &x| {
        (acc << 8) ^ (TABLE[((u32::from(x)) ^ (acc >> 24)) as usize])
    })
}
```
Both produce the same output (:+1:), but this PR is 10-15% slower.

For reference, I wrote:

```rust
lazy_static! {
    static ref TABLE: [u32; 256] = make_table(0x04c11db7, false);
}

pub fn calc_crc(data: &[u8]) -> u32 {
    !update(!0xefcb_f201, &TABLE, data, false)
}
```

File: Cargo.toml (hunk 1 above)
Comment (CLomanno) [reply Cargo.toml]: Removed, Criterion adds a lot of bloat.  I'll save the code for later in case it needs added back in

File: benches/bench.rs (hunk 2 above)
Comment (CLomanno) [reply benches/bench.rs]: removed

File: src/util.rs (hunk 4 above)
Comment (CLomanno) [reply src/util.rs]: fixed to `reflect_value_16` etc.

File: src/crc16.rs [hunk 6]
```diff
@@ -19,49 +34,124 @@ pub trait Hasher16 {
...
+///
+/// # Usage
+///
+/// call using Digest::write(&bytes)
+pub fn update(mut value: u16, table: &[u16; 256], bytes: &[u8], rfl: bool) -> u16 {
+    let shift = 8;
+
     for &i in bytes.iter() {
-        value = table[((value as u8) ^ i) as usize] ^ (value >> 8)
+        if true == rfl {
```
[the remark this answers is not in the document]
Comment (CLomanno) [reply src/crc16.rs]: updated to just `if rfl {`

File: src/crc16.rs [hunk 7]
```diff
@@ -19,49 +34,124 @@ pub trait Hasher16 {
...
 }
 
+/// Generates a generic USB 16 bit CRC (AKA CRC-16-IBM)
 pub fn checksum_usb(bytes: &[u8]) -> u16 {
-    return update(0, &USB_TABLE, bytes);
+    return update(0xFFFF, &USB_TABLE, bytes, true) ^ 0xFFFF;
 }
 
 impl Digest {
+    /// *Only works for reflected CRCs*
```
[the remark this answers is not in the document]
Comment (CLomanno) [reply src/crc16.rs]: You're right.  Fixed

**Comment by CLomanno** [comment]
@nickbabcock I like your speed improvement.  I'll work on putting it in

File: src/crc16.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (CLomanno) [reply src/crc16.rs]: It makes sense to take it out of the `update()` function.  I couldn't cleanly get it out of `make_table_crcXX()`

**Comment by mrhooray** [comment]
thanks @CLomanno @nickbabcock 
will update readme etc and release
